### DISCUSSION TITLE
How register liveness reducing is achieved

### DISCUSSION THREAD

**Discussion by wzhcz8902** [NVIDIA/cutlass#1515] [Q&A] [answered]
https://github.com/NVIDIA/cutlass/blob/5c447dd84f8ae0e1d48ff9a2eae26ce8c4958101/include/cutlass/gemm/threadblock/index_remat.h#L65-L69

Can you help me understand why it helps reduce register liveness as the comment says?

**Comment by hwu36** [comment, marked as answer]
instead of

```
mov R1, blockIdx.x
...
1000 lines of assembly
...
setp R1
```

we prefer to do
```
setp blockIdx.x
```

just recompute simple things every time we use it so that we can have more free registers.

**Comment by mrakgr** [comment]
Drat, I was hoping the Cuda compiler did that optimization on its own. I can think of many places in my ML library where I failed to apply it.

**Reply by mrakgr** [reply]
I am testing this optimization out on some simpler cases, and as far as I can tell, the Cuda compiler is very good at optimizing array indexing. I cannot see any difference in register liveness or the instruction count. The last operand of the IMAD instruction, which is a constant confusingly keeps changing between different versions, but I cannot see any benefit to applying this optimization by hand.

**Reply by FabioLuporini** [reply]
google made me land on this page. My 2 cents: perhaps it's just a legacy optimization (ie meaningful only with older compiler versions)
